docs/content/manual/manual.md

```diff
@@ -0,0 +1,5424 @@
+~~~
+add          # atomic (function call)
+[]           # iteration
+.posts       # indexing
+[0]?         # indexing
+.sections    # indexing
+[]           # iteration
+["title"]?   # indexing
+~~~
+
+We can transform this into an equivalent filter:
+
+~~~
+  add
+| .[]
+| .posts
+| .[0]?
+| .sections
+| .[]
+| .["title"]?
+~~~
+
+:::
+
+Filters inside a part of a compound path,
+such as `f` and `g` in `.[f][:g]`,
+are run with the _input given to the whole path_.
+
+::: Example
+
+When we run the filter `.arr[][.key]` on the input
+`{key: "a", arr: [{a: 1, b: 2}, {a: 3}]}`, then
+`.key` is run on the original input,
+not on the current value returned by `.arr[]`!
+To see the difference, let us first consider a **wrong** transformation:
+
+~~~
+  .arr          # --> [{a: 1, b: 2}, {a: 3}]
+| .[]           # -->  {a: 1, b: 2}, {a: 3}
+| .[.key]       # -->  error  (because .key is run with input {a: 1, b: 2} and yields null)
+~~~
+
+Now, let us consider a **correct** transformation:
+
+~~~
+  .key as $x    # --> "a"
+| .arr          # --> [{a: 1, b: 2}, {a: 3}]
+| .[]           # -->  {a: 1, b: 2}, {a: 3}
+| .[$x]         # -->  1, 3
+~~~
+
+:::
+
+::: Note
+Surprisingly, the filter `.[f]?` is **not** equivalent to `(.[f])?`.
+To see this, let us transform `.[f]?` to an equivalent filter like above:
+
+~~~
+  f as $x
+| .[$x]?
+~~~
+
+The difference shows when `f` causes an error --- in that case,
+ `.[f]?`  will raise the error, whereas
+`(.[f])?` will not raise _any_ error.
+:::
+
+::: Examples
+
+~~~
+.foo?
+{"foo": 42, "bar": "less interesting data"}
+42
+~~~
+
+~~~
+.foo?
+{"notfoo": true, "alsonotfoo": false}
+null
+~~~
+
+~~~
+.["foo"]?
+{"foo": 42}
+42
+~~~
+
+~~~
+[.foo?]
+[1,2]
+[]
+~~~
+
+:::
+
+
+# Arithmetic and Comparison
+
+We are now going to introduce operators for
+arithmetic (`+`, `-` `*`, `/`, `%`),
+equality (`==`, `!=`), and
+ordering (`<`, `<=`, `>`, `>=`).
+
+All operators in this section
+feed their input to both arguments and
+combine their results.
+This allows us to implement an averaging filter as `add / length` ---
+this feeds the input both to the `add` filter and the `length` filter,
+then performs the division of their results.
+
+Given two filters `f` and `g`, we can write
+`f + g`, `f == g`, `f < g` and so on
+to perform the desired operation on
+the outputs of the filters `f` and `g`.
+When `f` or `g` outputs multiple values, then
+all combinations of the operation are performed.
+
+::: Example
+Suppose that
+`f` outputs the values `1, 2` and
+`g` outputs the values `3, 4`.
+Then `f * g` outputs the values `3, 6, 4, 8`.
+:::
+
+::: Compatibility
+For any operator in this section such as `+`,
+in jq , `f + g` is equivalent to `g as $y | f as $x | $x + $y`, whereas
+in jaq, `f + g` is equivalent to `f as $x | g as $y | $x + $y`.
+That means that in the above example, jaq outputs the values
+`3, 4, 6, 8` instead of `3, 6, 4, 8`.
+Note that this difference shows only when both `f` and `g` produce multiple values.
```

Comment:
Oh god. I did not know about the behaviour of `[[1,2],[3,4]][0,1][0,1]`. That would make the description of compound paths a nightmare. Just checked, and gojq also does it like that. Is there a reason for this?

Reply:
I documented this behaviour in 4bbc505.

Reply:
Don't know if accidental or not. Think i talked to @nicowilliams about this on discord some months ago, also about the multi-eval behaviour. I will try find the transcript.

Reply:
Found this from discord:

![Screenshot 2024-11-06 at 13 32 17](https://github.com/user-attachments/assets/dd0589de-df0d-4f7b-a0ba-530cae05abd4)

... some confusion and debug cut away ...

![Screenshot 2024-11-06 at 13 31 50](https://github.com/user-attachments/assets/8e99fb79-0ba0-4662-8c3f-2d9f19ed7874)

Reply:
Thanks for posting the transcript. I think that this counts as a bug, because it's not actually necessary to evaluate `inputs` as often as it is done here.

Reply:
Yes I think so too and "fixing" it could only affect someone using input functions and multi index? i don't think there are other functions with side-effects? or could it affect error handling etc?

Reply:
I'm not familiar enough to talk about how jq handles this stuff, but I can think of a few other functions with side effects, like `halt` and `debug`. Error handling should not be affected, I think, but again, given my very limited knowledge of jq internals, I'm not very sure about this.